File: spring-boot-domain-persistence-modeling-java/src/main/java/x/scratch/parent/ParentRecord.java
package x.scratch.parent;

import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Table;
import x.scratch.UpsertableRecord;

@Builder(toBuilder = true)
@Data
@Table("parent")
public class ParentRecord implements ParentDetails,
        UpsertableRecord<ParentRecord> {
    @Id
    private Long id;
    private @NonNull String naturalId;
    private String value;
    private int version;

    static ParentRecord createRecordFor(final String naturalId) {
        return new ParentRecord(null, naturalId, null, 0);
    }

    @Override
    public ParentRecord updateWith(final ParentRecord upserted) {
        id = upserted.id;
        naturalId = upserted.naturalId;
        value = upserted.value;
        version = upserted.version;
        return this;
    }
}

File: spring-boot-domain-persistence-modeling-kotlin/src/main/kotlin/x/domainpersistencemodeling/java/ParentRecord.kt
package x.domainpersistencemodeling.java

import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table

@Table("parent")
data class ParentRecord(
        @Id var id: Long?,
        override var naturalId: String,
        override var value: String?,
        override var version: Int) : ParentDetails,
        UpsertableRecord<ParentRecord> {
    override fun updateWith(upserted: ParentRecord): ParentRecord {
        id = upserted.id
        naturalId = upserted.naturalId
        value = upserted.value
        version = upserted.version
        return this
    }

    companion object {
        internal fun createRecordFor(naturalId: String): ParentRecord {
            return ParentRecord(null, naturalId, null, 0)
        }
    }
}
